torch/csrc/distributed/c10d/NCCLUtils.cpp

```diff
@@ -26,9 +31,38 @@ ncclComm_t NCCLComm::getNcclComm() {
             ". ",
             commFailureMsg));
   }
+  // only wait for initialization if nonblocking mode is enabled
+  if (nccl_use_nonblocking()) {
+    waitUntilInitialized(nccl_nonblocking_timeout());
+  }
+
   return ncclComm_;
 }
 
+void NCCLComm::waitUntilInitialized(int timeoutSecs) {
+  auto startTimepoint = std::chrono::steady_clock::now();
+  while (!initialized_) {
+    auto currentTimepoint = std::chrono::steady_clock::now();
+    auto timeElapsed = std::chrono::duration_cast<std::chrono::seconds>(
+                           currentTimepoint - startTimepoint)
+                           .count();
+    if (timeElapsed > timeoutSecs) {
+      std::string err = "NCCL timeout in communicator initialization.";
+      TORCH_CHECK_WITH(DistBackendError, false, err);
+    }
+    if (ncclComm_) {
+      ncclResult_t result;
+      ncclCommGetAsyncError(ncclComm_, &result);
+      if (result == ncclSuccess) {
+        LOG(INFO) << "Rank " << rank_ << ": NCCL communicator is initialized.";
+        initialized_ = true;
+      }
+    }
```

Comment:
Strictly speaking, you'd need to put the `ncclCommGetAsyncError` check **BEFORE** the timeout check, to avoid a false positive timeout.